src/relation_handlers.py

```diff
@@ -418,17 +419,84 @@ def ready(self) -> bool:
         # TODO: True only when charm completes bootstrapping??
         return True
 
+    def can_service(self, event):
+        """Test if we can service the relation."""
+        return True
+
+    def update_broker_data(self, data, event):
+        """Update a broker response after it's been produced."""
+        pass
+
+    def get_key_params(self, event):
+        """Get the key name and the capabilities required."""
+        return event.client_app_name, None
+
     def _on_process_request(self, event):
+        if not self.can_service(event):
+            logger.info("Deferring handling of relation: %s" % self.relation_name)
+            event.defer()
+            return
+
         logger.info(f"Processing broker req {event.broker_req}")
         broker_result = process_requests(event.broker_req)
         logger.info(broker_result)
         unit_response_key = "broker-rsp-" + event.client_unit_name
         response = {unit_response_key: broker_result}
+        data = self.charm.get_ceph_info_from_configs(*self.get_key_params(event))
+        self.update_broker_data(data, event)
+
         self.interface.set_broker_response(
             event.relation_id,
             event.relation_name,
             event.broker_req_id,
             response,
-            self.charm.get_ceph_info_from_configs(event.client_app_name),
+            data,
         )
         # Ignore the callback function??
+
+
+class CephRadosGWProviderHandler(CephClientProviderHandler):
+    """Handler for the radosgw relation."""
+
+    def __init__(self, charm, callback_f):
+        super().__init__(charm, "radosgw", callback_f)
+        self.key_name = ""
+        self.force = False
+
+    @staticmethod
+    def _select_relation(relations, relation_id):
+        for relation in relations:
+            if relation.id == relation_id:
+                return relation
+
+    @staticmethod
+    def _remote_unit_name(client_name):
+        ridx = client_name.rindex("-")
+        return "ceph-radosgw/" + client_name[ridx + 1 :]  # noqa: E203
```

Comment:
nit: could be simplified as `return "ceph-radosgw/" + client_name.rsplit('-', 1)[1]`

Reply:
Done.